compiler/rustc_middle/src/middle/privacy.rs

```diff
@@ -103,31 +130,77 @@ impl<Id: Hash + Eq + Copy> AccessLevels<Id> {
         })
     }
 
-    pub fn set_access_level(&mut self, id: Id, tag: AccessLevel) {
-        let mut effective_vis = self.get_effective_vis(id).copied().unwrap_or_default();
+    pub fn get_effective_vis(&self, id: Id) -> Option<&EffectiveVisibility> {
+        self.map.get(&id)
+    }
+
+    pub fn iter(&self) -> impl Iterator<Item = (&Id, &EffectiveVisibility)> {
+        self.map.iter()
+    }
+
+    pub fn map_id<OutId: Hash + Eq + Copy>(&self, f: impl Fn(Id) -> OutId) -> AccessLevels<OutId> {
+        AccessLevels { map: self.map.iter().map(|(k, v)| (f(*k), *v)).collect() }
+    }
+
+    pub fn set_access_level(&mut self, id: Id, module_vis: Visibility, tag: AccessLevel) {
+        let mut effective_vis = self
+            .get_effective_vis(id)
+            .cloned()
+            .unwrap_or(EffectiveVisibility::from_vis(module_vis));
         for level in [
             AccessLevel::Public,
             AccessLevel::Exported,
             AccessLevel::Reachable,
             AccessLevel::ReachableFromImplTrait,
         ] {
             if level <= tag {
-                *effective_vis.get_mut(level) = Some(Visibility::Public);
+                *effective_vis.get_mut(level) = Visibility::Public;
             }
         }
         self.map.insert(id, effective_vis);
     }
 
-    pub fn get_effective_vis(&self, id: Id) -> Option<&EffectiveVisibility> {
-        self.map.get(&id)
-    }
-
-    pub fn iter(&self) -> impl Iterator<Item = (&Id, &EffectiveVisibility)> {
-        self.map.iter()
-    }
-
-    pub fn map_id<OutId: Hash + Eq + Copy>(&self, f: impl Fn(Id) -> OutId) -> AccessLevels<OutId> {
-        AccessLevels { map: self.map.iter().map(|(k, v)| (f(*k), *v)).collect() }
+    // parent_id is not necessarily a parent in source code tree, also
```

Comment:
I see what `parent_id` is not, but not what it is.
Could you say that it is the node from which the maximum effective visibility is inherited.